system/src/commonMain/kotlin/com/splendo/kaluga/system/network/state/NetworkStateRepo.kt

```diff
@@ -0,0 +1,123 @@
+/*
+ Copyright 2020 Splendo Consulting B.V. The Netherlands
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+
+package com.splendo.kaluga.system.network.state
+
+import co.touchlab.stately.concurrency.AtomicReference
+import com.splendo.kaluga.state.ColdStateRepo
+import com.splendo.kaluga.system.network.BaseNetworkManager
+import com.splendo.kaluga.system.network.Network
+import com.splendo.kaluga.system.network.unknownNetworkOf
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.launch
+
+class NetworkStateRepo(
+    private val networkManagerBuilder: BaseNetworkManager.Builder,
+) : ColdStateRepo<NetworkState>() {
+
+    interface Builder {
+        fun create(): NetworkStateRepo
+    }
+
+    private var _lastKnownNetwork = AtomicReference<Network>(Network.Unknown.WithoutLastNetwork(Network.Unknown.Reason.NOT_CLEAR))
```

Comment:
You never null this so it will leak (not a huge leak but still). To fix this we need some kind of dispose of the entire repo, I would rather not introduce that unless needed.

I think we can use `@ThreadLocal`, since `initialValue` and `deinitialize` should/could be same thread. (technically you can pass a multithreaded dispatcher to the ColdStateRepo but you currently block this in your own constructor)

Reply:
isn't `ThreadLocal` for singleton or top level declaration?

Reply:
Maybe I didn't understand what do you mean, but doing :
```
@ThreadLocal
internal var lastKnownNetwork: Network = Network.Unknown.WithoutLastNetwork(Network.Unknown.Reason.NOT_CLEAR)
```
I get
```
@ThreadLocal is applicable only to top level declarations
```

Reply:
Ah, interesting. Class is top level, but I guess it does not work for `class`, only for `object`

This solution is only for if it's need though, at which point does it fail as a regular var? Is it because the repo itself gets frozen?

Reply:
I can imagine the repo to be frozen at some point.

The real solution is that this variable should be in a local scope. I think we can achieve this when we use SharedFlow instead of BroadcastChannel at some later point.

The object that is potentially leaked is very small, let's just accept it for now.